express-notes-api/index.js

```diff
@@ -0,0 +1,109 @@
+const express = require('express');
+const app = express();
+const fs = require('fs');
+
+const arrayJson = [];
+
+const jsonMiddleware = express.json();
+app.use(jsonMiddleware);
+
+app.get('/api/notes', (req, res, next) => {
+  const data = require('./data.json');
+  const notesJson = data.notes; // object
+  for (const prop in notesJson) {
+    arrayJson.push(data.notes[prop]);
+  }
+  res.status(200).json(arrayJson);
+});
+
+app.get('/api/notes/:id', (req, res, next) => {
+  const data = require('./data.json');
+  const idNum = parseInt(req.params.id, 10);
+  if (idNum < 1 || !idNum) {
+    res.status(400).json({ error: 'id must be a positive integer' });
+  } else if (data.notes[idNum]) {
+    res.status(200).json(data.notes[idNum]);
+  } else {
+    res.status(404).json({ error: 'cannot find note with id ' + idNum });
+  }
+});
+
+app.post('/api/notes', (req, res, next) => {
+  const data = require('./data.json');
+  if (!req.body.content) {
+    res.status(400).json({ error: 'content is a required field' });
+    return;
+  }
+  let dataJson = data;
+
+  const newData = {
+    id: dataJson.nextId,
+    content: req.body.content
+  };
+
+  let nextId = dataJson.nextId;
+  dataJson.notes[nextId] = newData;
+  nextId++;
+  dataJson.nextId = nextId;
+  dataJson = JSON.stringify(dataJson, null, 2);
+
+  fs.writeFile('./data.json', dataJson, err => {
+    if (err) {
+      res.status(500).json({ error: 'An unexpected error occurred.' });
+    } else {
+      res.status(201).json(newData);
+    }
+  });
+});
+
+app.delete('/api/notes/:id', (req, res, next) => {
+  const data = require('./data.json');
+  let dataJson = data;
+
+  const idNum = parseInt(req.params.id, 10);
+  if (idNum < 1 || !idNum) {
+    res.status(400).json({ error: 'id must be a positive integer' });
+  } else if (data.notes[idNum]) {
+    delete data.notes[idNum];
+    dataJson = JSON.stringify(dataJson, null, 2);
+    fs.writeFile('./data.json', dataJson, err => {
+      if (err) {
+        res.status(500).json({ error: 'An unexpected error occurred.' });
+      }
+      res.sendStatus(204);
```

Comment:
Be sure to include `return` or `else` in here so that it is impossible for the server to try to respond twice and crash.